File: msp_frontend/src/models/Character.ts
import {Inventory} from "./Inventory";

/**
 * @class Class representing the player.
 */
export class Character {
  /**
   * @description The character's current health points.
   */
  private _healthPoints: number;

  /**
   * @description The character's maximum health points.
   */
  private _maxHealthPoints: number;

  /**
   * @description The character's strength attribute.
   */
  private _strength: number;

  /**
   * @description The character's luck attribute.
   */
  private _luck: number;

  /**
   * @description The character's charisma attribute.
   */
  private _charisma: number;

  /**
   * @description The amount of money the character has.
   */
  private _money: number;

  /**
   * @description The character's inventory.
   */
  private _inventory: Inventory;

  /**
   * @description Creates a new Character instance.
   * @param {number} healthPoints The character's starting health points.
   * @param {number} maxHealthPoints The character's maximum health points.
   * @param {number} strength The character's strength attribute.
   * @param {number} luck The character's luck attribute.
   * @param {number} charisma The character's charisma attribute.
   * @param {number} money The character's starting money.
   * @param {Inventory} inventory The character's inventory.
   */
  constructor(
    healthPoints: number,
    maxHealthPoints: number,
    strength: number,
    luck: number,
    charisma: number,
    money: number,
    inventory: Inventory
  ) {
    this._healthPoints = healthPoints;
    this._maxHealthPoints = maxHealthPoints;
    this._strength = strength;
    this._luck = luck;
    this._charisma = charisma;
    this._money = money;
    this._inventory = inventory;
  }

  /**
   * @description Gets the character's current health points.
   * @returns The character's current health points.
   */
  get healthPoints(): number {
    return this._healthPoints;
  }

  /**
   * @description Sets the character's current health points.
   * @param value The new health points for the character.
   */
  set healthPoints(value: number) {
    this._healthPoints = value;
  }

  /**
   * @description Gets the character's maximum health points.
   * @returns The character's maximum health points.
   */
  get maxHealthPoints(): number {
    return this._maxHealthPoints;
  }

  /**
   * @description Sets the character's maximum health points.
   * @param value The new maximum health points for the character.
   */
  set maxHealthPoints(value: number) {
    this._maxHealthPoints = value;
  }

  /**
   * @description Gets the character's strength attribute.
   * @returns The character's strength attribute.
   */
  get strength(): number {
    return this._strength;
  }

  /**
   * @description Sets the character's strength attribute.
   * @param value The new strength value for the character.
   */
  set strength(value: number) {
    this._strength = value;
  }

  /**
   * @description Gets the character's luck attribute.
   * @returns The character's luck attribute.
   */
  get luck(): number {
    return this._luck;
  }

  /**
   * @description Sets the character's luck attribute.
   * @param value The new luck value for the character.
   */
  set luck(value: number) {
    this._luck = value;
  }

  /**
   * @description Gets the character's charisma attribute.
   * @returns The character's charisma attribute.
   */
  get charisma(): number {
    return this._charisma;
  }

  /**
   * @description Sets the character's charisma attribute.
   * @param value The new charisma value for the character.
   */
  set charisma(value: number) {
    this._charisma = value;
  }

  /**
   * @description Gets the character's current amount of money.
   * @returns The character's current amount of money.
   */
  get money(): number {
    return this._money;
  }

  /**
   * @description Sets the character's charisma attribute.
   * @param value The new charisma value for the character.
   */
  set money(value: number) {
    this._money = value;
  }

  /**
   * @description Gets the character's charisma attribute.
   * @returns The character's charisma attribute.
   */
  get inventory(): Inventory {
    return this._inventory;
  }

  /**
   * @description Sets the character's charisma attribute.
   * @param value The new charisma value for the character.
   */
  set inventory(value: Inventory) {
    this._inventory = value;
  }
}

File: MSP_Backend/src/main/java/de/szut/msp_backend/models/character/Character.java
package de.szut.msp_backend.models.character;

import de.szut.msp_backend.exceptions.ItemNotFoundException;
import de.szut.msp_backend.models.inventory.Inventory;
import de.szut.msp_backend.models.item.Consumable;
import de.szut.msp_backend.models.item.GenericItem;
import lombok.AllArgsConstructor;
import lombok.Data;
import org.springframework.transaction.annotation.Transactional;

@Data
@AllArgsConstructor
public class Character
{
    private int healthPoints;
    //maxHealthPoints might need additional balancing
    private int maxHealthPoints;
    private int strength;
    private int luck;
    private int charisma;
    //TODO charisma hier bitte als Wert zwischen 0 bis 10 halten
    private int money;
    private Inventory inventory;

    /**
     * Money that gets added to the player's money.
     *
     * @param addMoney the amount that gets added
     */
    public void addMoney(int addMoney)
    {
        this.money = money + addMoney;
    }

    /**
     * Removes money from the player
     *
     * @param subMoney the money amount that gets removed
     */
    public void removeMoney(int subMoney)
    {
        this.money = money - subMoney;
    }

    /**
     * This method adds healtpoints by eating.
     *
     * @param consumable the consumable with a healthpoint value that would add to the players healthpoints
     * @return the healpoints added to the player's health, not higher than maxHealthpoints.
     */
    public int eat(Consumable consumable)
    {
        if (maxHealthPoints > healthPoints + consumable.getHealthGain())
        {
            healthPoints = healthPoints + consumable.getHealthGain();
            return healthPoints;
        }
        healthPoints = maxHealthPoints;
        return healthPoints;
    }

    /**
     * Player can sell an item to a trader.
     *
     * @param item  the item that gets sold.
     * @param price the money the player gets for selling.
     * @throws ItemNotFoundException if the item is not found.
     */
    @Transactional
    public void sellItemToTrader(GenericItem item, int price) throws ItemNotFoundException
    {
        removeItemFromInventory(item, 1);
        addMoney(price);
    }

    /**
     * An item can be bought from a trader by the player with money,
     * if the inventory has space for the item and the player has enough money.
     *
     * @param item  the item the trader sells
     * @param price the costs for the item
     * @return BuyItemResponse for different occasions (NOTENOUGHMONEY, NOTENOUGHSPACE, OK)
     */
    @Transactional
    public Enum<BuyItemResponse> buyItemFromTrader(GenericItem item, int price)
    {
        if (this.money < price)
        {
            return BuyItemResponse.NOT_ENOUGH_MONEY;
        }
        if (!inventory.isNotFull())
        {
            return BuyItemResponse.NOT_ENOUGH_SPACE;
        }
        removeMoney(price);
        addItemToInventory(item, 1);
        return BuyItemResponse.OK;
    }

    /**
     * Adds an item to the player's inventory.
     *
     * @param item   the item that gets added
     * @param amount the amount of the added item.
     */
    public void addItemToInventory(GenericItem item, int amount)
    {
        inventory.addItem(item, amount);
    }

    /**
     * Removes chosen item from player's inventory.
     *
     * @param item   the item from the inventory that gets removed.
     * @param amount the amount of the item that gets removed,
     *               if the amount is equal or less than the item amount in the inventory.
     * @throws ItemNotFoundException if the item is not in the inventory.
     */
    public void removeItemFromInventory(GenericItem item, int amount) throws ItemNotFoundException
    {
        if (!inventory.isItemPresent(item) || (inventory.isItemPresent(item) && inventory.getItems().get(item) < amount))
        {
            throw new ItemNotFoundException();
        }
        inventory.removeItem(item, amount);
    }

    /**
     * clears the inventory.
     */
    public void clearInventory()
    {
        inventory.clearInventory();
    }
}
